File: js/pattern.js
// Access pattern detector (Adaptive-prefetch-spec §4).

export const SEQUENTIAL_THRESHOLD = 10;
export const RANDOM_THRESHOLD = 100;
export const HISTORY_SIZE = 32;
export const MIN_OBSERVATIONS = 8;
export const UPGRADE_SEQUENTIAL_THRESHOLD = 100;

export const PATTERN_UNKNOWN = "unknown";
export const PATTERN_SEQUENTIAL = "sequential";
export const PATTERN_RANDOM = "random";
export const PATTERN_MIXED = "mixed";

/** Observes record(index) / seek calls and classifies access patterns. */
export class AccessPatternDetector {
  constructor() {
    /** @type {number[]} */
    this.accesses = Array(HISTORY_SIZE).fill(-1);
    this.writePos = 0;
    this.filled = 0;
    this._sequentialRuns = 0;
    this._randomJumps = 0;
    this.lastIndex = -1;
  }

  sequentialRuns() {
    return this._sequentialRuns;
  }

  getLastIndex() {
    return this.lastIndex;
  }

  /** @param {number} index */
  observe(index) {
    const idx = index;
    if (this.lastIndex >= 0) {
      const delta = Math.abs(idx - this.lastIndex);
      if (delta <= SEQUENTIAL_THRESHOLD) {
        this._sequentialRuns = Math.min(this._sequentialRuns + 1, 0xffffffff);
      } else if (delta > RANDOM_THRESHOLD) {
        this._randomJumps = Math.min(this._randomJumps + 1, 0xffffffff);
      }
    }
    this.accesses[this.writePos] = idx;
    this.writePos = (this.writePos + 1) % HISTORY_SIZE;
    if (this.filled < HISTORY_SIZE) this.filled++;
    this.lastIndex = idx;
  }

  /** @returns {string} */
  pattern() {
    const total = this._sequentialRuns + this._randomJumps;
    if (total < MIN_OBSERVATIONS) return PATTERN_UNKNOWN;
    if (this._sequentialRuns > this._randomJumps * 3) return PATTERN_SEQUENTIAL;
    if (this._randomJumps > this._sequentialRuns) return PATTERN_RANDOM;
    return PATTERN_MIXED;
  }

  /**
   * Predicted next record indices when pattern is sequential (§4.4).
   * @param {number} depth
   * @param {number} recordCount
   * @returns {number[]}
   */
  predictNext(depth, recordCount) {
    if (this.pattern() !== PATTERN_SEQUENTIAL || this.lastIndex < 0) return [];
    const start = this.lastIndex + 1;
    const out = [];
    for (let i = 0; i < depth; i++) {
      const idx = start + i;
      if (idx < recordCount) out.push(idx);
    }
    return out;
  }
}

File: py/pattern.py
"""Access pattern detector (Adaptive-prefetch-spec §4)."""
from __future__ import annotations

SEQUENTIAL_THRESHOLD = 10
RANDOM_THRESHOLD = 100
HISTORY_SIZE = 32
MIN_OBSERVATIONS = 8
UPGRADE_SEQUENTIAL_THRESHOLD = 100

PATTERN_UNKNOWN = "unknown"
PATTERN_SEQUENTIAL = "sequential"
PATTERN_RANDOM = "random"
PATTERN_MIXED = "mixed"


class AccessPatternDetector:
    """Observes record(index) calls and classifies access patterns."""

    __slots__ = (
        "accesses",
        "write_pos",
        "filled",
        "_sequential_runs",
        "_random_jumps",
        "last_index",
    )

    def __init__(self) -> None:
        self.accesses: list[int] = [-1] * HISTORY_SIZE
        self.write_pos = 0
        self.filled = 0
        self._sequential_runs = 0
        self._random_jumps = 0
        self.last_index = -1

    def sequential_runs(self) -> int:
        return self._sequential_runs

    def observe(self, index: int) -> None:
        idx = index
        if self.last_index >= 0:
            delta = abs(idx - self.last_index)
            if delta <= SEQUENTIAL_THRESHOLD:
                self._sequential_runs = min(self._sequential_runs + 1, 0xFFFFFFFF)
            elif delta > RANDOM_THRESHOLD:
                self._random_jumps = min(self._random_jumps + 1, 0xFFFFFFFF)
        self.accesses[self.write_pos] = idx
        self.write_pos = (self.write_pos + 1) % HISTORY_SIZE
        if self.filled < HISTORY_SIZE:
            self.filled += 1
        self.last_index = idx

    def pattern(self) -> str:
        total = self._sequential_runs + self._random_jumps
        if total < MIN_OBSERVATIONS:
            return PATTERN_UNKNOWN
        if self._sequential_runs > self._random_jumps * 3:
            return PATTERN_SEQUENTIAL
        if self._random_jumps > self._sequential_runs:
            return PATTERN_RANDOM
        return PATTERN_MIXED

    def predict_next(self, depth: int, record_count: int) -> list[int]:
        """Predicted next record indices when pattern is sequential (§4.4)."""
        if self.pattern() != PATTERN_SEQUENTIAL or self.last_index < 0:
            return []
        start = self.last_index + 1
        out: list[int] = []
        for i in range(depth):
            idx = start + i
            if idx < record_count:
                out.append(idx)
        return out
